### PR TITLE
(2) Quinn Introduction Chapter, Data transfer

### PR DESCRIPTION
**Pull request by TimonPost**
This PR is part of a series of PR's regarding the book creation as described in #865. Each page of this chapter will have its own PR for reducing PR size.

- Network Introduction (#917)
- QUINN Introduction (#868)
    - Certificate Configuration (#920)
   - Connection Setup (#921)
    **- Data Transfer (#922)**

## Live Version

A live version of this page can be found [here](https://timonpost.github.io/quinn/). This book shows all the work that I have up-to-now. The pages that are visible but don't have a PR yet should are welcome to have general feedback.

## Chapter Motivation

This chapter will go through the protocol by examples. The chapters are in the required order, first, we look at configuring a certificate which is **required**, then at setting up a connection, and finally at sending data about this connection. In the future, more pages can be added.

After reading this chapter the reader should be able to **send data** over a secured with a real or self-signed **certificate**, or insecure, **connection**.

The target audience is people who are new to the library and want to get started quickly by some examples and a short explanation. It is not meant to describe the entire API. 

## Page Motivation

The page introduces the reader to sending data over the endpoint configured in the page (#921). After reading the user is familiar with bidirectional, unidirectional streams, and datagram messaging.

## Remarks

- Some links to other chapters don't work yet, but will be working as soon other PR's are merged.

### CODE REVIEW

File: docs/book/src/quinn/data-transfer.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
+# Data Transfer
+
+In the [previous chapter](set-up-connection.md) we characterized how to set up an [Endpoint][Endpoint]
+and then get access to a [NewConnection][NewConnection] instance.
+Now we will continue with the subject of sending data over this connection.
+
+## Multiplexing
+
+A QUIC stream can be compared to a TCP stream, eventhough if it is based on UDP.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:9]:
```suggestion
A QUIC stream can be compared to a TCP stream, even though if it is based on UDP. 
```

File: docs/book/src/quinn/data-transfer.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+In the [previous chapter](set-up-connection.md) we characterized how to set up an [Endpoint][Endpoint]
+and then get access to a [NewConnection][NewConnection] instance.
+Now we will continue with the subject of sending data over this connection.
+
+## Multiplexing
+
+A QUIC stream can be compared to a TCP stream, eventhough if it is based on UDP. 
+The big difference is that you are not limited to a single stream. 
+You can open multiple streams between two peers, this is also called multiplexing.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:11]: This doesn't seem accurate. "multiplexing" is the act of combining data from multiple streams in a single stream, which isn't really what you're referring to here.

File: docs/book/src/quinn/data-transfer.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+Stream multiplexing can have a significant positive effect on application performance if the resources allocated to streams are properly prioritized.
+Currently multiplexing it is used in a protocol like HTTP/2, but unlike HTTP multiplexing, QUIC does not automatically multiplex.
+This is an advantage because the user gets full control over the multiplexing. 
+Another difference compared to a TCP stream is that QUIC is not limited to one kind of stream but can open both bidirectional and unidirectional streams.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:16]: I think TCP sockets can close one side of the connection, which also results in a unidirectional stream? So this is not quite accurate. I'd just leave the TCP comparison out here and briefly explain that there are four kinds of streams for each QUIC connection: client/server-initiated bi/uni-directional.

File: docs/book/src/quinn/data-transfer.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+Now we will continue with the subject of sending data over this connection.
+
+## Multiplexing
+
+A QUIC stream can be compared to a TCP stream, eventhough if it is based on UDP. 
+The big difference is that you are not limited to a single stream. 
+You can open multiple streams between two peers, this is also called multiplexing.
+
+Stream multiplexing can have a significant positive effect on application performance if the resources allocated to streams are properly prioritized.
+Currently multiplexing it is used in a protocol like HTTP/2, but unlike HTTP multiplexing, QUIC does not automatically multiplex.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:14]: This comparison doesn't make much sense to me. HTTP isn't really a stream-oriented protocol, it's a request-response protocol.

File: docs/book/src/quinn/data-transfer.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+## Stream Types
+
+Quinn offers three ways to send your data. 
+Two stream-based and one message-based.
+
+| Type | Description | Reference |
+| :----- | :----- | :----- |
+| **Bidirectional Stream** | two way stream communication. | see [open_bi][open_bi] |
+| **Unidirectional Stream** | one way stream communication. | see [open_uni][open_uni] |
+| **Unreliable Messaging** | message based unreliable communication. | see [send_datagram][send_datagram] |
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:27]: Maybe should mention that datagrams are an extension, not yet standard, and not required to be implemented.

File: docs/book/src/quinn/data-transfer.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+async fn iterate_streams(mut connection: NewConnection) -> anyhow::Result<()> {
+    // Iterate unidirectional streams with only the receiving side.
+    while let Some(Ok(recv)) = connection.uni_streams.next().await { }
+    // Iterate bidirectional streams with both sent and receiving side.
+    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await { }
+    // Iterate arrived datagrams.
+    while let Some(Ok(bytes)) = connection.datagrams.next().await { }
+
+    Ok(())
+}
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:48]: "opened" doesn't seem like the best term here, let's use "incoming" instead to make it clear that these were initiated by the peer?

File: docs/book/src/quinn/data-transfer.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        open_bi()
+        .await?;
+
+    Ok(())
+}
+```
+
+## Bidirectional Streams
+
+With bidirectional streams you can carry data in both directions, for example, from the initiator to the peer and in reverse.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:71]:
```suggestion
With bidirectional streams data can be sent in both directions, for example, from the connection initiator to the peer and the other way around.
```

File: docs/book/src/quinn/data-transfer.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        println!("{:?}", recv.read_to_end(50).await?);
+    }
+
+    Ok(())
+}
+```
+
+## Unreliable Messaging
+
+With unreliable messaging you can transfer data unreliable over bare UDP.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:140]: This isn't really "over bare UDP"; datagram frames are embedded in QUIC packets, which are bundled up in UDP datagrams.

File: docs/book/src/quinn/data-transfer.md (hunk 3 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:16]: Good one!

File: docs/book/src/quinn/data-transfer.md (hunk 2 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:11]: Agree, I will rephrase as you posted.

File: docs/book/src/quinn/data-transfer.md (hunk 4 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:14]: I am trying to shed light on the comparison between how multiplexing works in HTTP protocol and in quic. In quic it is not done automatically were as in HTTP it, right?

Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:14]: I am going to rewrite this part a bit differently. I think there is some unnecessary info here.

File: docs/book/src/quinn/data-transfer.md (hunk 6 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:48]: I will just remove this word and leave it at 'iterate over various streams'. Incoming is may not be correct because it can be initiated from the endpoint itself.

File: docs/book/src/quinn/data-transfer.md (hunk 8 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:140]: Good point. I will mention that instead.

**Comment by TimonPost** [comment]
Thanks for the review. I think it is ready for another review. I removed a huge amount of text under 'multiplexing' and rephrased some sentences with the provided feedback.

File: docs/book/src/quinn/data-transfer.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
+# Data Transfer
+
+In the [previous chapter](set-up-connection.md) we characterized how to set up an [Endpoint][Endpoint]
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:3]: `characterizing` doesn't really make sense in this sentence.

File: docs/book/src/quinn/data-transfer.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+and then get access to a [NewConnection][NewConnection] instance.
+Now we will continue with the subject of sending data over this connection.
+
+## Multiplexing
+
+Multiplexing is the act of combining data from multiple streams into a single stream. 
+This can have a significant positive effect on the performance of the application. 
+QUIC it itself does not perform multiplexing instead it gives the programmer full control over this process.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:11]: This sentence doesn't make sense to me. QUIC does multiplex frames for all streams into what is conceptually a single stream of QUIC packets.

File: docs/book/src/quinn/data-transfer.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    // Iterate bidirectional streams with both sent and receiving side.
+    while let Some(Ok((sent, recv))) = connection.bi_streams.next().await { }
+    // Iterate arrived datagrams.
+    while let Some(Ok(bytes)) = connection.datagrams.next().await { }
+
+    Ok(())
+}
+```
+
+*Open various kinds streams*
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:44]:
```suggestion
*Open various kinds of streams*
```

File: docs/book/src/quinn/data-transfer.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+| **Bidirectional Stream** | two way stream communication. | see [open_bi][open_bi] |
+| **Unidirectional Stream** | one way stream communication. | see [open_uni][open_uni] |
+| **Unreliable Messaging (extension)** | message based unreliable communication. | see [send_datagram][send_datagram] |
+
+## How to Use
+
+New streams can be created with the methods [open_bi][open_bi], [open_uni][open_uni] of type [Connection][Connection].
+An instance of this type, together with existing streams, can be found in the [connection][connection] field of [NewConnection].
+
+*Iterate over various streams*
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:29]: Here and below, I'm not sure `iterate` makes sense to me. What we're really doing here is waiting for incoming streams, which is more like "receiving" than "iterating", perhaps? Also the example code can be misleading. Also maybe the example should maybe make clear that you probably shouldn't use code like this because it would never get to receive bidirectional streams or datagrams, since it'd be waiting for unidirectional streams forever (until the peer closes its connection).

File: docs/book/src/quinn/data-transfer.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        send.finish().await?;
+    }
+
+    Ok(())
+}
+```
+
+## Unidirectional Streams 
+
+With unidirectional streams, you can carry data only in one direction, for example, from the initiator of the stream to its peer.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:101]:
```suggestion
With unidirectional streams, you can carry data only in one direction: from the initiator of the stream to its peer.
```

File: docs/book/src/quinn/data-transfer.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        println!("{:?}", recv.read_to_end(50).await?);
+    }
+
+    Ok(())
+}
+```
+
+## Unreliable Messaging
+
+With unreliable messaging you can transfer data unreliable UDP.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:133]: This isn't really a sentence. Also maybe clarify why unreliability is desirable for some kinds of data, and why datagrams can be a better solution rather than having lots of short unidirectional streams.

File: docs/book/src/quinn/data-transfer.md (hunk 10 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:11]: I mean, like in HTTP/2 multiplexing is performed for you, in QUIC you're responsible for it.

Comment (djc) [reply docs/book/src/quinn/data-transfer.md:11]: I don't agree. In the Quinn API, the user is certainly not on the hook to merge together parts of different streams into particular packets.

Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:11]: That's right. Maybe I need to rephrase it to the user being in control over how to allocate streams and spread data over them.

File: docs/book/src/quinn/data-transfer.md (hunk 12 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:29]: Good one. I mainly added this example to show a very minimal usage of receiving from different streams. I will add a note for that. Also receiving would make more sense yea.

File: docs/book/src/quinn/data-transfer.md (hunk 14 above)
Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md:133]: What is the benefit for many unidirectional streams vs unreliable messaging?

**Comment by TimonPost** [comment]
It is ready for an other review!

File: docs/book/src/quinn/data-transfer.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
+# Data Transfer
+
+In the [previous chapter](set-up-connection.md) it is explained how to set up an [Endpoint][Endpoint]
+and then get access to a [NewConnection][NewConnection] instance.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:4]:
```suggestion
The [previous chapter](set-up-connection.md) explained how to set up an [Endpoint][Endpoint]
and then get access to a [NewConnection][NewConnection] instance.
```

File: docs/book/src/quinn/data-transfer.md [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        println!("{:?}", recv.read_to_end(50).await?);
+    }
+
+    Ok(())
+}
+```
+
+## Unreliable Messaging
+
+With unreliable messaging you can transfer data over UDP without reliability.
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:134]:
```suggestion
With unreliable messaging you can transfer data without reliability. 
```

File: docs/book/src/quinn/data-transfer.md [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    Ok(())
+}
+```
+
+## Unreliable Messaging
+
+With unreliable messaging you can transfer data over UDP without reliability. 
+This could be useful if data-arrival isnt essential or when needing of high-throughput matters and reliab
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:135]: Unfinished sentence?

```suggestion
This could be useful if data arrival isn't essential or when needing of high throughput matters and reliab
```

File: docs/book/src/quinn/data-transfer.md [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+
+*send datagram*
+
+```rust
+async fn sent_unreliable(connection: Connection)-> anyhow::Result<()> {
+    connection.
+        send_datagram(b"test".into())
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:142]:
```suggestion
    connection
        .send_datagram(b"test".into())
```

File: docs/book/src/quinn/data-transfer.md [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+    Ok(())
+}
+```
+
+*iterating datagram stream(s)*
+
+```rust
+async fn receive_datagram(mut connection: NewConnection) -> anyhow::Result<()> {
+    while let Some(Ok(receivedBytes)) = connection.datagrams.next().await {
```
Comment (djc) [inline docs/book/src/quinn/data-transfer.md:153]:
```suggestion
    while let Some(Ok(received_bytes)) = connection.datagrams.next().await {
```

**Comment by TimonPost** [comment]
Okay, I updated the changes and added some other ones as well.

File: docs/book/src/quinn/data-transfer.md (hunk 14 above)
Comment (djc) [reply docs/book/src/quinn/data-transfer.md]: Many unidirectional streams get you reliability without ordering (so no head-of-line blocking).

Comment (djc) [reply docs/book/src/quinn/data-transfer.md]: Since this is still open, I'm waiting for you to follow up, but perhaps you are waiting for my review? Just realized that it's the end of January and we haven't completely merged your work yet, sorry about that!

Comment (TimonPost) [reply docs/book/src/quinn/data-transfer.md]: This is an old comment and some things have changed already.  Tho I will introduce 'Many unidirectional streams get you reliability without ordering (so no head-of-line blocking)' into this text.

**Comment by TimonPost** [comment]
I think I addressed all your comments. It's ready for any further process.

**Comment by TimonPost** [comment]
Are there any remarks left?

**Review by djc — APPROVED** [review]
